File: 0133-clone-graph/0133-clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    void deepClone(Node node, Node res, Node[] nodes) {
        if(node == null || nodes[node.val] != null) return;
        nodes[node.val] = res;
        for(Node n : node.neighbors) {
            if(res.neighbors.isEmpty()) res.neighbors = new ArrayList<>();
            if(nodes[n.val] == null) {
                res.neighbors.add(new Node(n.val));
                deepClone(n, res.neighbors.get(res.neighbors.size() - 1), nodes);
            } else {
                res.neighbors.add(nodes[n.val]);
            }
        } 
    }

    public Node cloneGraph(Node node) {
        if(node == null) return null;
        Node res = new Node(node.val);
        deepClone(node, res, new Node[101]);
        return res;
    }
}

File: 0133-clone-graph/0133-clone-graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    void deepClone(Node* node, Node* ans, unordered_map<int,Node*> &presentNodes){
        if(node == NULL || presentNodes.find(node->val) != presentNodes.end()) return;
        
        presentNodes[node->val] = ans;
        
        for(Node* adjNode: node->neighbors){
            if(ans->neighbors.empty()) ans->neighbors = vector<Node*>();
            
            if(presentNodes.find(adjNode->val) == presentNodes.end()){
                ans->neighbors.push_back(new Node(adjNode->val));
                
                deepClone(adjNode, ans->neighbors.back(), presentNodes);
            }else{
                ans->neighbors.push_back(presentNodes[adjNode->val]);
            }
            
        }
        
    }
    Node* cloneGraph(Node* node) {
        if(node == NULL) return node;
        
        Node* ans = new Node(node->val);
        
        unordered_map<int,Node*> presentNodes;
        
        deepClone(node, ans, presentNodes);
        
        return ans;
    }
};